### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by uronito** [SwiCago/HeatPump#11]
Hi al, kayno.

I'm Albertoo from nicegear.nz. Can you help me?

If i try add more mqtt_publish in hpSeetingsChanges the sketch dont' work (only reads temperature) not send commands to heatpump or mqtt /heatpump.

The code i use is:

void hpSettingsChanged() {
    const size_t bufferSize = JSON_OBJECT_SIZE(6);
    const size_t bufferSizeDomoticz_power = JSON_OBJECT_SIZE(3);
    const size_t bufferSizeDomoticz_mode = JSON_OBJECT_SIZE(3);
    const size_t bufferSizeDomoticz_fan = JSON_OBJECT_SIZE(3);
    const size_t bufferSizeDomoticz_temperature = JSON_OBJECT_SIZE(3);
    
    
    DynamicJsonBuffer jsonBuffer(bufferSize);
    
    DynamicJsonBuffer jsonBufferDomoticz_power(bufferSizeDomoticz_power);
    DynamicJsonBuffer jsonBufferDomoticz_mode(bufferSizeDomoticz_mode);
    DynamicJsonBuffer jsonBufferDomoticz_fan(bufferSizeDomoticz_fan);
    DynamicJsonBuffer jsonBufferDomoticz_temperature(bufferSizeDomoticz_temperature);
    
    JsonObject& root = jsonBuffer.createObject();
    JsonObject& domoticz_power = jsonBufferDomoticz_power.createObject();
    JsonObject& domoticz_mode = jsonBufferDomoticz_mode.createObject();
    JsonObject& domoticz_fan = jsonBufferDomoticz_fan.createObject();
    JsonObject& domoticz_target_temp = jsonBufferDomoticz_temperature.createObject();

    
    heatpumpSettings currentSettings = hp.getSettings();
    
    root["power"]       = currentSettings.power;
    root["mode"]        = currentSettings.mode;
    root["temperature"] = currentSettings.temperature;
    root["fan"]         = currentSettings.fan;
    root["vane"]        = currentSettings.vane;
    root["wideVane"]    = currentSettings.wideVane;
    char buffer[512];
    root.printTo(buffer, sizeof(buffer));
    bool retain = true;
    mqtt_client.publish(heatpump_topic, buffer, retain);
    domoticz_power["command"] = "setuservariable";
    domoticz_power["idx"]     = 13;
    domoticz_power["value"]   = currentSettings.power;
    char buffer_domoticz_power[512];
    domoticz_power.printTo(buffer_domoticz_power, sizeof(buffer_domoticz_power));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz_power);
    domoticz_mode["command"] = "setuservariable";
    domoticz_mode["idx"]     = 12;
    domoticz_mode["value"]   = currentSettings.mode;
    char buffer_domoticz_mode[512];
    domoticz_mode.printTo(buffer_domoticz_mode, sizeof(buffer_domoticz_mode));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz_mode);
    domoticz_fan["command"] = "setuservariable";
    domoticz_fan["idx"]     = 15;
    domoticz_fan["value"]   = currentSettings.fan;
    char buffer_domoticz_fan[512];
    domoticz_fan.printTo(buffer_domoticz_fan, sizeof(buffer_domoticz_fan));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz_fan);
    domoticz_target_temp["command"] = "setuservariable";
    domoticz_target_temp["idx"]     = 14;
    domoticz_target_temp["value"]   = currentSettings.temperature;
    char buffer_domoticz_target_temp[512];
    domoticz_target_temp.printTo(buffer_domoticz_target_temp, sizeof(buffer_domoticz_target_temp));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz_target_temp);

}

why cant't publlish mqtt messages ?

Thanks in advance!!!!!!!

**Comment by uronito** [comment]
I already have it solved :),

One question, it is normal that when sending an order to the heat pump, in the heatpump mqtt topic comes out the repeated command twice.

Thanks in advance !!!!!

topic /heatpump:
{"power":"OFF","mode":"FAN","temperature":21,"fan":"1","vane":"1","wideVane":"|"}
{"power":"OFF","mode":"FAN","temperature":21,"fan":"1","vane":"1","wideVane":"<<"}

**Comment by kayno** [comment]
@uronito - how did you solve the problem?

regarding seeing that topic change twice - I am seeing this too. I think it's happening because the wideVane setting is not working - when you update wideVane the callback is called, but then the heatpump  changes back to the last known wideVane setting because it was invalid, so the callback fires again.

We need to work out the wideVane codes, which is being covered in #7.

**Comment by SwiCago** [comment]
@uronito exactly what @kayno  said, wideVane is at the moment unknown. As far as i recall, it did not work for hadley either and is on the list of things to discover. I know for a fact it is not an option for kumo cloud device, so i wonder if it can even be set???

**Comment by uronito** [comment]
Hi Kayno ,Al,

@kayno, here is my code:

void hpSettingsChanged() {
    const size_t bufferSize = JSON_OBJECT_SIZE(6);
    const size_t bufferSizeDomoticz = JSON_OBJECT_SIZE(3);
    DynamicJsonBuffer jsonBuffer(bufferSize);
    JsonObject& root = jsonBuffer.createObject();
    DynamicJsonBuffer jsonBufferDomoticz(bufferSizeDomoticz);
    JsonObject& domoticz = jsonBufferDomoticz.createObject();
    heatpumpSettings currentSettings = hp.getSettings();
    root["power"]       = currentSettings.power;
    root["mode"]        = currentSettings.mode;
    root["temperature"] = currentSettings.temperature;
    root["fan"]         = currentSettings.fan;
    root["vane"]        = currentSettings.vane;
    root["wideVane"]    = currentSettings.wideVane;
    char buffer[512];
    root.printTo(buffer, sizeof(buffer));
    bool retain = true;
    mqtt_client.publish(heatpump_topic, buffer, retain);
    domoticz["command"] = "setuservariable";
    domoticz["idx"]     = 13;
    domoticz["value"]   = currentSettings.power;
    char buffer_domoticz[512];
    domoticz.printTo(buffer_domoticz, sizeof(buffer_domoticz));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz, retain);
    domoticz["command"] = "setuservariable";
    domoticz["idx"]     = 12;
    domoticz["value"]   = currentSettings.mode;
    
    domoticz.printTo(buffer_domoticz, sizeof(buffer_domoticz));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz, retain);
    domoticz["command"] = "setuservariable";
    domoticz["idx"]     = 15;
    domoticz["value"]   = currentSettings.fan;
    
    domoticz.printTo(buffer_domoticz, sizeof(buffer_domoticz));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz, retain);
    domoticz["command"] = "setuservariable";
    domoticz["idx"]     = 14;
    domoticz["value"]   = String(currentSettings.temperature);
    
    domoticz.printTo(buffer_domoticz, sizeof(buffer_domoticz));
    mqtt_client.publish(domoticz_mqtt, buffer_domoticz, retain );

}

@kayno, @all , 

topic /heatpump:
{"power":"OFF","mode":"FAN","temperature":21,"fan":"1","vane":"1","wideVane":"|"}
{"power":"OFF","mode":"FAN","temperature":21,"fan":"1","vane":"1","wideVane":"<<"}

my heat pump dont have wideVame, always i send power or mode or temperature or fan in the heatpump mqtt topic comes out the repeated command twice.

**Comment by uronito** [comment]
i Have original mac-557-ife mitsubishi wifi adaptor for melcloud, but dont have wide vane in melcloud app.

Thanks

**Comment by kayno** [comment]
re the double publish - it will happen even though you are not setting it. It might actually be related to two issues - #7 and #9:

#7 because we haven't discovered how to control the wideVane
#9 is that there is a default wantedSetting for wideVane, and each time it gets pushed the heatpump (firing the callback) the heatpump then flicks it back to whatever it originally was, firing the callback again

does your heatpump have a horizontal vane (left/right swing) at all? or just a vertical (up/down vane) vane? This is where future versions of the library might be able to detect the heatpump type or what features it has and adjust accordingly.

@uronito - which version of the code are you running? do you have my commit from ~12hrs ago?

**Comment by uronito** [comment]
@kayno ,

i have a ducted pead rp100 jaa, dont have vane,widevane.
I use last swicago version 23 hours ago.
I havent see your last version. Tomorrow i will probe it.
I just saw that the latest version of melcloud if you have to control widevane.
Thanks @kayno, @SwiCago

**Comment by SwiCago** [comment]
@uronito kumo cloud doesnt have wideVane. If as you say melcloud does, please to a tcp dump and see if it talks to their web in plain http. New kumo talks https, so no way to se what payload it gets from server. If melcloud is http, then you can get dump the payload as it receives it. It should be an xml, as i have seen posted somewhere else...and in that xml os literally the hex code that the device sends the heatpump.

**Comment by SwiCago** [comment]
@kayno, maybe we should default wanted wideVane as soon as it comes in from current wideVane, at least until we figure it out

**Comment by SwiCago** [comment]
@kayno @uronito default wideaVane for HPs that don't have wideVane should be 0x00, so what ever that is represented in the map.

**Comment by uronito** [comment]
@SwiCago ,

I try to get a tcp dump of the data coming from melcloud but I did not get it, I think they are encrypted.

@kayno 

i had probe now your last version.The callback  fires again. I think that default position for ducted heatpumps is (|) not (<<). I will try remap widevane array.

Thanks!!!

**Comment by uronito** [comment]
Wowwwww!!!!

At least!!!!!!! 

I have managed to not trigger the callback for units that do not have widevane

The error was here:

HeatPump.cpp:

  HeatpumpSettings defaultSettings = {POWER_MAP [0], MODE_MAP [3], TEMP_MAP [9], FAN_MAP [2], VANE_MAP [1], WIDEVANE_MAP [2]};

I've changed it to:

  HeatpumpSettings defaultSettings = {POWER_MAP [0], MODE_MAP [3], TEMP_MAP [9], FAN_MAP [2], VANE_MAP [1], WIDEVANE_MAP [0]};

And so it is always kept in the pump and in the code the same value for wide vane with which the callback is not triggered.

**Comment by kayno** [comment]
@uronito glad to hear that worked! In my latest commits, I have removed the `wantedSettings` defaults completely, and left it to be populated the first time settings are received from the heatpump. I am assuming your heatpump will send a `0x01` (which is what is in `WIDEVANE_MAP[0]`) and set it in wantedSettings as the default, which will achieve the same as you have done, but also remove the problem raised in #9.

Can you please test my fork and see if this is the case? https://github.com/kayno/HeatPump

I have been testing it and it seems fine, but it would be nice to have someone else try before I submit a pull request to push it up to @SwiCago's repo.

**Comment by uronito** [comment]
I'll prove it right away and I'll tell you, Kay, give me 10 minutes

**Comment by uronito** [comment]
@kayno last version give me a error

exit status 1
'class HeatPump' has no member named 'setPacketReceivedCallback'

**Comment by kayno** [comment]
sorry @uronito, forgot to update the example in the library, doing that now

**Comment by uronito** [comment]
Thanks @kayno :)

**Comment by kayno** [comment]
try now :)

**Comment by uronito** [comment]
now works perfect!!!!!!

but if i set  '{"power":"OFF"}'  heatpump tell me {"power":"OFF","mode":"HEAT","temperature":28,"fan":"AUTO","vane":"AUTO","wideVane":"<<"}

and my last seetings in my heat pump before was

{"power":"OFF","mode":"HEAT","temperature":24,"fan":"AUTO","vane":"AUTO","wideVane":"<<"}

**Comment by uronito** [comment]
I think it is a normal behavior of the heat pump. When it detects that a device is connected by the port cn105, the heat pump sends that package with the adjustments by deefcto that came from factory, in this case mode heat with 28º. It only happens the first time, so I think it's normal for the heat pump, I think with melcloud it was also happening. So I think the code is perfect

**Comment by kayno** [comment]
ok interesting. I haven't seen it occur on my heatpump, so maybe a model thing. if you wanted to dig further, you could enable debug by sending a message to heatpump/debug/set with a payload of just the string "on", and then subscribe to /heatpump/debug to see all the packets going back and forth. There is a lot of output in debug mode though, so it can take a while to debug!!

thanks for the feedback on the code. i will sort a pull request :)

**Comment by uronito** [comment]
@kayno ,

Glad to help, I'll take a look at debug mode to see if I find out anything.

Many thanks to you for your efforts

Regards :)

**Comment by uronito** [comment]
Could you try something?

Disconnect the pump from the power supply.

wait a few seconds

Connect to power.

And send the command off by mqtt

See if it happens to you about the temperature.

Thank you

**Comment by kayno** [comment]
I will have to try tomorrow, it's near midnight here!

**Comment by uronito** [comment]
Ok Kayno.

Thankssss!!!!  Goodnight

**Comment by uronito** [comment]
I have been testing Kayay during the day and the truth is that it works quite well. Just one thing, I noticed that if I change fast mode with the remote mqtt does not detect the change, there must be a pause of 2 seconds at least for the code to detect the change and show it in the heatpump / set theme

Thanks

**Comment by kayno** [comment]
Is "fast mode" a button on your remote?

**Comment by uronito** [comment]
No, jejeje

I mean if I change fast, that is, by pressing the mode key quickly

**Comment by kayno** [comment]
Oh ok, quick fingers!!

If you do press mode fast, does it always report on mqtt the final mode you set? Or does that get missed?

We may need to review the `delay()` calls in `update()`

**Comment by uronito** [comment]
@kayno always no, missed presses.

if presses are in 1-2  second , missed some changes.

tomorrow i will try adjust delay.

Regards

**Comment by SwiCago** [comment]
@uronito  @kayno , great team work guys. And thanks for adding to the compatability list. When we have a bigger list, I will make a compatability list file. This library is really taking off. I honestly didn't think anyone would else would need it, but am very glad I shared it to github

**Comment by uronito** [comment]
@kayno , @swicago, Equally, it is a pleasure to be able to help you to get this project forward :)

@kayno, today i will probe your last version, I saw you did it a few minutes ago.

Thanks @kayno, @swicago

**Comment by kayno** [comment]
@uronito, would be great if you could test it, to make sure it's not just working on my heatpump! thanks :)

**Comment by uronito** [comment]
@kayno , i try your last version.

Works like a charm!!!!.

Regards

**Comment by uronito** [comment]
@kayno, 

Too fast i said it.

There is a problem.

Whenever I turn on the pump in heat mode per mqtt, it always turns on 28º, ie what in the previous version passed only the first time (default values of the pump) now happens whenever I send a command per mqtt.

**Comment by uronito** [comment]
if i change 


  hp.setSettingsChangedCallback(hpSettingsChanged);
  hp.setRoomTempChangedCallback(sendCurrentRoomTemperature);
  hp.setPacketCallback(hpPacketDebug);
  hp.connect(&Serial);

per

   hp.connect(&Serial);
  hp.setSettingsChangedCallback(hpSettingsChanged);
  hp.setRoomTempChangedCallback(sendCurrentRoomTemperature);
  hp.setPacketCallback(hpPacketDebug);

works perfect.

Regards

**Comment by uronito** [comment]
Can you probe the same?

Thanks!!!!

**Comment by kayno** [comment]
@uronito, can you try these steps with @SwiCago's repo  (in other words without my latest changes)?

I am unable to replicate, so I want to see if its just my changes that are causing this.

**Comment by uronito** [comment]
I have try swicago's repo last version.

Te same problem. 

Until you set the temperature from mqtt. Whenever you turn on or off the pump by mqtt {"power": "ON"} or {"power": "OFF"} the pump will use the default temperatures, in this case for heat will be 28 °

Regards

**Comment by uronito** [comment]
I think the problem is here:

if(!wantedSettings) {
            wantedSettings = receivedSettings;
          }
i think that the firs time the heatpump connects, wantesssetings dont' set to receivedsettings because 28º is max of my heatpump (0x00) in the temperature array.

Regards

**Comment by uronito** [comment]
Wooooowwww!!!!

With this modification: (heatpump.cpp)
-------if(!wantedSettings) {--------> delete this
wantedSettings = receivedSettings;
--------}----------> delte this

Now works perfect always in swicago's code and kayno's code

Regards!!!!

**Comment by SwiCago** [comment]
@uronito get latest merge, make the change. @kayno can you test his change, i am not using the mqtt feature yet. When confirmed we can do a pull merge again

**Comment by uronito** [comment]
@swicago, i have made modification.

Regards!!!!

**Comment by SwiCago** [comment]
@uronito mergeed. @kayno  please confirm that it works as expected

**Comment by SwiCago** [comment]
@kayno, there was a reason you checked if it was null, now with this merge wanted will always become current. @uronito i am not sure that is the correct logic, as my wanted settings may not be what the heat pump is set as. In example i set heat via esp, but ir remote sets it to fan. Via esp i still want heat, not fan. It should execute a call back that values are different from wanted and then it is up to the controlling software to decide if it should change it or not. However i do agree that if the wanted and current differ, that the hp library should only notify of the change once. Better yet would be to creat a prior current settings and compare that with current settings, if it differs notify. Wanted settings should not be changed automatically by library

**Comment by uronito** [comment]
@swicago,

I've been testing the latest version of Kayno with the modification I've done before, and the truth is that it works perfectly. Whenever I want to make a change by esp makes it perfect and the esp perfectly stores the modifications that are made from the pump.

Regards

**Comment by SwiCago** [comment]
@uronito @kayno , I understand we want heatpump to store current settings, but current settings should never override wanted settings.
Here is how I see it. 
ESP connects
ESP sets its wantedSettings
ESP send update command
ESP receives updateConfirm
ESP receives callBack, because current settings != received settings (received stored in current)
IR makes change
ESP receives callBack, because current settings != received settings (received stored in current)
It then update the receiving end to decide to leave as is, or send update commend, which send the wanted settings it originally wanted...unless updated by code in mean time.
Wanted settings should never be overwritten by current or received ever!

I just don't see why wantedSettings would ever be changed by library, user sketch decides it's values.
If user wants to be in sync, maybe get currrentSettings in sketch and update wanted to it, via sketch as setup. Library should not do this.

urinto, can you remove the wantedSettings = receivedSettings completely and see if it still works for you.
Also, make sure on your sketch to be sure to get currentSettings once and set to your wanted settings or at least once set all wantedSettings, before you allow MQTT to do shcnages. I think that is where the confusion comes. When update is called, all settings must be sent to heatpump, it cannot just send one change. I believe Kayno set the wantedSettings to receivedSettings for first run, so it at least has a defaulted value. However IMO it is upto the sketch to get the settings and set all wanted settings at least once prior to calling update and then there after individual settings can be made. 
So either kayno's change to set wanted at least once goes back in and requires more testing or your sketch has to set settings at least once.

**Comment by kayno** [comment]
@SwiCago - to answer your question about why i checked if `wantedSettings` is null and then set it if it is, as you alluded to this is done so that on the very first run (and only the first run) I believe that `wantedSettings` should be synced to `currentSettings`. Otherwise, it has to be set somehow else, such as:

- the library sets arbitrary "default" values to `wantedSettings` (which is previously did) however that puts `wantedSettings` out of sync with what the heatpump is doing, and calling an individual `setSomeSettingSettings()` (e.g. `setPowerSetting()`) and then `update()` will see all settings changed (we fixed this in #9)
- or as you suggested the sketch calling the library has to do it. I don't like this idea, as this is a one-off scenario when the heatpump first powers on (or the ESP is first connected) and this little complexity should be abstracted from the "user" (where user is the person using the library to write a sketch). I think that a user should be able to `connect()`, `setPowerSetting()` and `update()` without having to make sure other settings are in sync (they would have to do a `getSettings()` and `setSettings()` in tandem which just doesn't make sense!

Personally I think the original idea (i.e. the solution from #9) is what we want. Going back to @uronito's earlier post:

> I think the problem is here:
> 
> if(!wantedSettings) {
> wantedSettings = receivedSettings;
> }
> i think that the firs time the heatpump connects, wantesssetings dont' set to receivedsettings because 28º is max of my heatpump (0x00) in the temperature map.

@uronito - can you explain this some more? 28 is in `TEMP` and maps to `0x03` in `TEMP_MAP`, yet you are saving you are getting `0x00` (which maps to 31)?

it would be great @uronito to see the debug topic when your heatpump starts. The MQTT example now has [`_debugMode` set to true](https://github.com/SwiCago/HeatPump/blob/master/examples/mitsubishi_heatpump_mqtt_esp8266/mitsubishi_heatpump_mqtt_esp8266.ino#L16), so it starts up in debug mode and you can see the `connect()` packets, and also the very first settings packet that comes in. That might help us to debug what the heatpump is reporting and what `wantedSettings` should be. If you open a terminal and subscribe to the topic `heatpump/debug`, then turn the heatpump on and you will see all the output. Post the first 10-20 odd lines here and we should be able to see what we need!

I also would like to test that the [`operator!` function](https://github.com/SwiCago/HeatPump/blob/master/src/HeatPump.cpp#L41) for `heatpumpSettings` is working correctly, and confirm that `wantedSettings` is only being set to `currentSettings` on the very first boot. Maybe my logic is wrong here??

Finally, going back to what I said earlier about abstracting complexities from the user... I wonder if there is a case to make `update()` private, and call it from the library every time one of the `setSettings()` functions is called? I was thinking about this from the point of view that if we were documenting each function of the library, for each of the `setSettings()` functions (e.g. `setPowerSetting()`, `setModeSetting()`, etc) we would keep repeating "Remember, this does not make the change, it just prepares things for you to call `update()`. It is when you call update() that the setting is actually changed". The very name of these functions - `setSettings()`, `setPowerSettings()`, etc is already implying that it will "set the setting", so why not make it happen then and there? Doing this might help us with some of the complexity of keeping `wantedSettings` in sync. If the IR remote for my heatpump had a button on it that said "update" or "change" or "send to heatpump" that you had to press after pressing the mode or temp buttons then I might think otherwise, but it doesn't :)

**Comment by SwiCago** [comment]
@kayno, the reason for update was for webbased application where more than one setting is set. 
The IR remote send all settings at once, not just one. So if we make update private, then sync will have to see if things changed and update. Also it would have to make sure that all individual settings have completed before update, so there is no race condition. Lastly if it is private, no way of knowing it is success like now on packet 61...i agree, wanted should be set at first boot as current settings of HP. But still think update should be done after user program determines it should be sent to HP.

**Comment by kayno** [comment]
@swicago, so are you saying #16 should be reversed? if so, I agree. We then need to look at @uronito's debug output to see what is coming in and what `wantedSettings` is being set to.

with the `update()` call being private, I hadn't thought of that angle (web based app). If we are saying that sketches that use the library should mimic the remote and send all settings, then we probably need to remove the individual set functions (e.g. `setPowerSetting()`) as the remote never does this...  However I don't think we should remove the individual set functions, as I don't think we are trying to mimic the remote...

> Also it would have to make sure that all individual settings have completed before update, so there is no race condition.

How would this race occur? you would still send all settings at once if just `setPowerSetting()` was called and it called `update()` itself. It uses `wantedSettings` to achieve this, and all the other settings (other than power) will be whatever the `currentSetting` are....

And if you want to set all at once, as in the web based app, you have `setSettings()` to do this...

The return value for update() can simply be returned by the set function that is being called, so that's not a problem.

**Comment by SwiCago** [comment]
@kayno I reverted the last commit. I think it is just an issue where @uronito  may have not set all his settings pr some other issue. He will have to check tge debug as you suggested. 
Basically i want to give people the choice to use individual settings or all at once, maybe we should allow the end user to choose if update is done automatically or not. Maybe a race could happen if a setting is set and hp updates and at same time another setting is set causing another update action. You dont press your remote a million times either lol...i only built the individual getters and setters, because for some it is easier to read than doing one set and one get for the whole thing. If we want to automaticially update as you suggest, then i think the set functions should set a last internal update time and sync checks to see of a given time has passed, where we are sure the user app finished setting all settings, whether it is all at once or individually. 10ms should be enough for that

**Comment by SwiCago** [comment]
@kayno  how about we add a new method
void autoUpdate(boolean bool) 
Default value is false
Then on each set function update a time variable
Then on sync do the following
If(autoUpdate && lastChange >10ms) {update() }

**Comment by kayno** [comment]
@SwiCago let's create a new issue to work through auto update. your idea to make a new method `autoUpdate` is one I'd like to think through more. I'd advocate for default value to be true, as the default method should always be the easy option. I have some more thoughts, but lets get a separate issue so that we can leave this one to sort the issue @uronito is having!

thanks for the revert of #16 too.

**Comment by SwiCago** [comment]
@kayno, you cant set default tto true, otherwise you run the risk of doing an update against the will of an end user app, before they set it to false. 
Just like on a db, you dont default auto commit to true either

**Comment by kayno** [comment]
i hear you about commits in db land, but i think this is different. Different in the sense that if you are calling a function `heatpump.setPowerSetting("ON")`, I think anyone would expect that to switch on the heatpump. 

the library never calls a set function itself (apart from setSettings() which calls all the individual sets, but setSettings() is never called by the library), so there is no risk of doing an update against a user's will - they choose when to call update()/set functions/sync()/etc...

thinking about this from the other way, what is the advantage for the end user to be able to call any of the set functions, and then update()? what would they want to do in between?

**Comment by kayno** [comment]
@uronito we can debug the issue with temp 28/31 that you are having in this issue. The other issue is for auto update, which is separate.

Are you able to post your debug log here, from the debug topic, as I requested above?

**Comment by uronito** [comment]
@kayno give me a minutes and post it.

Regards

**Comment by kayno** [comment]
great. given your different temperature range, it might be good if you could also set the temperature to a few different settings and show us the debug output when doing so.

**Comment by uronito** [comment]
```
{"packetRecv":"fc` 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 08 03 00 00 00 00 00 00 00 00 00 4f "}

```

**Comment by uronito** [comment]
there is for:

mode:heat, power:off, temperature:23, fan:2, vane:auto, wide:auto

**Comment by kayno** [comment]
that looks fine - temp 23 maps to the 11th byte - 0x08 - as per what we have in the code.

I am interested in seeing the 1st 10 lines of debug when you disconnect/reconnect the heatpump from the power supply.

Also id like to see the debug when you press use the IR remote to set the temperature to the highest temperature, and the lowest.

**Comment by uronito** [comment]
@kayno,

I dont have ir remote, i have wall molunted remote (par21maa)

I had to add a delay of 80 seconds before sending anything to the pump since my pump is slow to start. These that I sent to you are the first packages that I receive in the topic heatpump / debug.

Regards

**Comment by uronito** [comment]
I'm going to try to lower the delay

**Comment by uronito** [comment]
with 20 secods of delay dont work.

I'm going to try to higher the delay

**Comment by uronito** [comment]
with 40 seconds dont connect.

I'm going to try to higher the delay, 60 seconds

**Comment by uronito** [comment]
```
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
{"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
{"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
{"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}

```
Those are the first lines that i receive in the haatpump/debug

**Comment by kayno** [comment]
@uronito can you confirm that you have _debugMode set to true in the mqtt example sketch? You should see the connect() packets first, like this:

``
heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "} // first connect() packet
heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "} // second connect() packet
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "} // sync() packet
heatpump/debug {"packetRecv":"fc 7a 01 30 01 00 54 00 38 ec fe 3f d8 04 10 40 c0 87 fe 3f 00 00 "} // first response
heatpump/temperature {"roomTemperature":0}
```

also you will need to set the callbacks before calling connect() - I think you may have changed that when debugging. Might be best to grab a fresh copy of the example and library from the repo and try. we need to see the connect() packets.

finally, can you subscribe to the topic `heatpump/#` (not `heatpump/debug`). Sorry, that was my mistake - I gave you the wrong topic. `heatpump/#` will show us everything coming and going from the heatpump, just like in my example above (e.g. it also shows `heatpump/temperature'). The command I use in linux is:

```
mosquitto_sub -h localhost -p 1883 -v -t 'heatpump/#' -u <MQTT_USERNAME> -P "<MQTT_PASSWORD>"
```

Leave out the -u and -P options if you do not need to authenticate to your MQTT broker.

**Comment by SwiCago** [comment]
@uronito @kayno, I see his HP returning on first temp 0x09 which maps to roomTemperature being 19°C
Are you using getTemperature or getRoomTemperature? when getting call back setRoomTempChangedCallback....The correct one to use would be getRoomTemperature

**Comment by uronito** [comment]
i probe it.

give me a minutes.

Regards

**Comment by uronito** [comment]
```
root@raspberrypi:~# mosquitto_sub -h localhost -p 1883 -v -t 'heatpump/#'
heatpump/temperature {"roomTemperature":0}
heatpump {"power":"OFF","mode":"HEAT","temperature":25,"fan":"2","vane":"AUTO","wideVane":"<<"}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump {"power":"OFF","mode":"HEAT","temperature":25,"fan":"2","vane":"AUTO","wideVane":"<<"}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/temperature {"roomTemperature":19}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 00 01 06 03 00 00 00 00 00 00 00 00 00 51 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 09 00 00 00 00 00 00 00 00 00 00 00 00 51 "}

```

**Comment by uronito** [comment]
i don't see connect packets

**Comment by kayno** [comment]
@uronito and the setcallback functions are being called before connect in your sketch?

**Comment by uronito** [comment]
```
  hp.setSettingsChangedCallback(hpSettingsChanged);
  hp.setRoomTempChangedCallback(sendCurrentRoomTemperature);
  hp.setPacketCallback(hpPacketDebug);
  delay(65000);
  hp.connect(&Serial);
```

**Comment by uronito** [comment]
is correct?

**Comment by kayno** [comment]
Very odd. The initial request packets are missing too. I'm unsure now how to continue!

**Comment by uronito** [comment]
i think the connect packets are sent because if i lower the delay the heatpump don't comunicate whit esp8266

**Comment by kayno** [comment]
Yeah, so they should be coming through to the debug topic. Unless it's failing to publish to mqtt.

**Comment by uronito** [comment]
yo creo que es eso, por alguna razon no estan siendo pùblicados

**Comment by uronito** [comment]
I have been testing several things but I can not see the connection packets.

I do not know what may be failing. I have tried different configurations of mqtt and I never get the connection packets

**Comment by SwiCago** [comment]
@uronito, @schotime pointed out that the struct wantedSettings is not actually null, so the if(!wantedSetttings) is incorrect and should be if(!wantedSettings.power) as power would definitely be null. Maybe that is the issue you are having? Or the other issue may be that your thermostat on cn51, conflicts with data on cn105 with esp. Maybe you can only have one or the other controller???

**Comment by SwiCago** [comment]
@kayno @uronito , I finally got Mosquito MQTT up and running and tested kayno's sketch. My on first time turning ON, temperature in current settings changed to 31 as reported by uronito. However, if I set temperature once via MQTT it stays and never changes again when turning HP on/off. So somehow wanted settings is not be set right on first boot. I have a feeling the struct may auto default it's values.

**Comment by uronito** [comment]
That's just what happened to my @swicago.

With the modification I made, it always worked correctly, always marked the temperature stored in the pump when sending by mqtt power: ON.

I can not see the connection packets through mqtt debug. I only see all the packages that come later, but I can not see the initials. Why can it be @swicago?

**Comment by SwiCago** [comment]
@uronito I see my connect packet sent, but not the response from connect. There is an if statement checking checksum, maybe it is not valid??? Or maybe we miss a message between connect/setup and when sketch finally goes to loop/sync. I will add a sync, right after connect and see what that yields
@kayno I do not think the struct is NULL at start, @schotime pointed that out earlier, I think it is all set to lowest values, in temp case 0x00, which is 31. I will do some more debugging and see what I can do, just sucks to re-upload code.

**Comment by uronito** [comment]
@swicago, can not be what you say because in the first round equals wantedssetings to currentsettings so it should carry the same value that the pump has stored so if it is set to (0x00) it is because that value is empty and take wantedsetting["temp"] position 0 in the map. I gave him a thousand turns and it has to be that

**Comment by SwiCago** [comment]
@uronito @kayno I added a sync() in the connect() right after the packets are sent. Here is my dump

heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "}
heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 7a 01 30 01 00 54 3f 80 ec fe 3f c4 eb fe 3f 98 87 fe 3f 00 00 "}
heatpump/debug {"packetRecv":"fc 7a 01 30 01 00 54 00 e0 fc fe 3f 70 ec fe 3f 00 54 00 00 00 00 "}
heatpump/temperature {"roomTemperature":0}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 07 00 00 00 00 00 03 be 00 00 00 00 92 "}
heatpump {"power":"ON","mode":"FAN","temperature":31,"fan":"AUTO","vane":"AUTO","wideVane":"|"}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 0c 00 01 ad ad fe 00 00 00 00 00 00 00 f5 "}
heatpump/temperature {"roomTemperature":22}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 07 00 00 00 00 00 03 be 00 00 00 00 92 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7a "}
heatpump/debug {"packetRecv":"fc 62 01 30 10 03 00 00 0c 00 01 ad ad fe 00 00 00 00 00 00 00 f5 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}

**Comment by uronito** [comment]
```
heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "}
heatpump/debug {"packetSent":"fc 5a 01 30 02 ca 01 a8 "}
heatpump/debug {"packetSent":"fc 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7b "}
heatpump/debug {"packetRecv":"fc 7a 01 30 01 00 54 3f 80 ec fe 3f c4 eb fe 3f 98 87 fe 3f 00 00 "}
heatpump/debug {"packetRecv":"fc 7a 01 30 01 00 54 00 e0 fc fe 3f 70 ec fe 3f 00 54 00 00 00 00 "}
```
I dont see this. Everything else I see

That was happening to me. The first round did not fit well, once adjusted by mqtt everything worked correctly.

That's why I made the modification. With the modification even the first time went right

Regards

**Comment by uronito** [comment]
```
void HeatPump::connect(HardwareSerial *serial) {
  _HardSerial = serial;
  _HardSerial->begin(2400, SERIAL_8E1);
  
  // settle before we start sending packets
  delay(2000);

  // send the CONNECT packet twice - need to copy the CONNECT packet locally
  byte packet[CONNECT_LEN];
  memcpy(packet, CONNECT, CONNECT_LEN);
  for(int count = 0; count < 2; count++) {
    writePacket(packet, CONNECT_LEN);
    delay(1100);
  }
  sync();
}
```

@swicago, your modification is is like this?

**Comment by SwiCago** [comment]
@uronito I just called sync, since first time sync is called it is RQST_PKT_SETTINGS, second time would be RCVD_PKT_ROOM_TEMP

**Comment by uronito** [comment]
I still do not see the packets that are sent from the connection and the connection packets that the pump returns

**Comment by SwiCago** [comment]
@uronito try pulling down my dev branch and try it.

**Comment by SwiCago** [comment]
Uch, having some commit issue I cant resolve. I need to delete my dev branch and re-enter it

**Comment by SwiCago** [comment]
@uronito please try my dev branch, got it uploaded

**Comment by uronito** [comment]
i'm going to probe it

**Comment by SwiCago** [comment]
@uronito I created a firstRun boolean, instead of looking at wantedSettings

**Comment by SwiCago** [comment]
@uronito you may have to add that connect delay, I have just been pulling power from ESP, not HP to test it.

**Comment by SwiCago** [comment]
@uronito , @kayno Oh, something just dawned on my. I remember when I was probing the KUMO cloud packets, I saw all values set to 0x00 when setting POWER to OFF. When KUMO cloud set POWER back on, it also sent it's wanted Temperature. I wonder if that is the case with uronito's themostat controller? If that is the case, we may have to ignore receivedSettings Temp on POWER = OFF

**Comment by uronito** [comment]
I still do not see my connection packets and the settings stored in the pump still do not work the first time.

**Comment by uronito** [comment]
I still think that wantedsettings is not matched to receivedssettings the first time the pump is started

**Comment by SwiCago** [comment]
Did you get my dev copy? I added a variable called firstBoot .

**Comment by SwiCago** [comment]
look at the constructor and tell me if you see the following 
HeatPump::HeatPump() {
  lastSend = 0;
  infoMode = false;
  autoUpdate = false;
  firstRun = true;
}

**Comment by uronito** [comment]
yes, i have probe your dev last version.

**Comment by SwiCago** [comment]
and you do 
mosquitto_sub -h localhost -p 1883 -v -t 'heatpump/#'

**Comment by uronito** [comment]
yes

**Comment by SwiCago** [comment]
maybe your heatpump doesn't return a connect packet? Is your hp older? I believe these packets say what they are, model, serial firmware version

**Comment by uronito** [comment]
The problem is that I do not see the connection packets that I send to the pump. But I know that those packets are sent because if under the delay of 65 seconds, the connection to the pump is not established.

My bomb is from 2010

**Comment by SwiCago** [comment]
Last question, do you start with everything OFF or ON? If OFF, I believe your other controller is setting everything to 0x00

**Comment by SwiCago** [comment]
My KUMO cloud did this when turning the unit OFF

**Comment by uronito** [comment]
I do not understand the question.

But with the change I made the other day, it works perfect from the start. With the last position memorized in the pump from the first turn

**Comment by SwiCago** [comment]
I just updated the library again, please try it. Make sure to over write your local copy and restart IDE

**Comment by uronito** [comment]
always do it.

I'm going to probe it

**Comment by SwiCago** [comment]
In this one, I only currentSettings.temperature if mode = ON

**Comment by SwiCago** [comment]
er POWER = ON

**Comment by uronito** [comment]
23 minutes ago?

**Comment by SwiCago** [comment]
it was seconds ago, but looking at the code it is what I just uploaded 
          if(receivedSettings.power == "ON") { 
            receivedSettings.temperature = lookupByteMapValue(TEMP_MAP, TEMP, 16, data[5]); }
          else { 
            receivedSettings.temperature = currentSettings.temperature; 
          }

**Comment by SwiCago** [comment]
@uronito, now it says 4min ago lol...guess github is not in sync

**Comment by uronito** [comment]
I do not think it works swicago, because also happens to me that if I start by mqtt {power: ON} the pump is set to 28 (max) in my pump (0x00) and if under the temperature by the thermostat of the wall (par21maa) and I leave it set to 24, if I then turn off the pump by mqtt {power: OFF} the pump goes off but at 28.

Can you probe with your ir remote?

**Comment by SwiCago** [comment]
My IR, doesn't override any of the ESP settings, so I am not experiencing this issue with my prior run

**Comment by uronito** [comment]
That happens because until the temperature is adjusted by mqtt, wantedsettigns is empty

**Comment by uronito** [comment]
When I make changes with the esp I see them reflected in my wall thermostat and on the contrary also

**Comment by SwiCago** [comment]
It is supposed to be set by that first Sync to get current settings. An it is to be set with what the HP is returning. In my case it never update wantedSettings as in the old if(!wantedSettings), this if never happens, the struct is never NULL, just the lowest values available.
I think @kayno will have to help again. I just don't understand how you do not get firstBoot to override your wantedSettings with what the HP things it has.

**Comment by SwiCago** [comment]
@kayno @uronito On my wall HP I get the following under MODE bytes for the follwoing when changing via IR
AUTO = 0x08
HEAT = 0x0b
COOL = 0x09
DRY = 0x0a 
FAN = 0x07
So I guess some settings have more than one valid value

**Comment by SwiCago** [comment]
@kayno, @uronito I also noticed that on my IR remote, if I set temp Fahrenheit, every other value gets returned as 31C
ex: 69F = 21C, 70F = 31C
What I do see is another byte being set when I change temperature byte 17 
so byte 11 and 17 change depending on temperature setting.a2 for 63F
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 0b 0e 02 01 00 00 03 a2 00 00 00 00 99 "}
heatpump {"power":"ON","mode":"HEAT","temperature":17,"fan":"1","vane":"1","wideVane":"|"}
a1 for 62F
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 0b 1f 02 01 00 00 03 a1 00 00 00 00 89 "}
heatpump {"power":"ON","mode":"HEAT","temperature":31,"fan":"1","vane":"1","wideVane":"|"}
a0 for 61F
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 0b 0f 02 01 00 00 03 a0 00 00 00 00 9a "}
heatpump {"power":"ON","mode":"HEAT","temperature":16,"fan":"1","vane":"1","wideVane":"|"}

as said some translate to 31C for our byte as it cannot be found in our map
Example 74F  byte 11 = 0x18, byte 17 = 0xaf... 0x18 does not map, so it gets defaulted to 31.
heatpump/debug {"packetRecv":"fc 62 01 30 10 02 00 00 01 09 18 02 01 00 00 03 af 00 00 00 00 84 "}
heatpump {"power":"ON","mode":"COOL","temperature":31,"fan":"1","vane":"1","wideVane":"|"}

I also, changed my MODE sets to what was returned by HP
AUTO = 0x08
HEAT = 0x0b
COOL = 0x09
DRY = 0x0a
FAN = 0x07
But they mostly get ignored and nothing is returned. I think I will need to hook up the KUMO cloud and sniff the RX after each change.

**Comment by schotime** [comment]
This seems like exactly what was happening to me. First connect it would set the would not set the wanted settings so the update would be different for the auto update and set everything to the defaults and since the temp was out of range it was set to 31. 

However my Dev PR which I'm sure @SwiCago copied correctly did rectify it for me. It assumes that a sync happens first which it should now.

**Comment by kayno** [comment]
woah - trying to get up to speed with all of this discussion!

Thinking about why !wantedSetttings wasn't working, and you had to look at !wantedSettings.power. The ! operator for the heatpumpSettings struct is:

```
bool operator!(const heatpumpSettings& settings) {
  return !settings.power && 
         !settings.mode && 
         !settings.temperature && 
         !settings.fan &&
         !settings.vane &&
         !settings.wideVane;
}
```

it may be that when we declare wantedSettings in Heatpump.h:

```
heatpumpSettings wantedSettings;
```

That the values of each property are set to garbage, and not 0/null/etc. I'm basing this on what I read here:  http://stackoverflow.com/questions/29336359/how-do-you-check-if-a-struct-is-initialized-in-c

A fix might be to add the following to the HeatPump constructor:

```
wantedSettings = {};
```

however I am not home right now and cannot test this. I will be able to do some testing tonight, in about 10hrs time!

The bool firstRun flag might be a better solution, but we still should be able to test for !wantedSettings - just have to get the code right :)

**Comment by kayno** [comment]
Alternatively, we could add a property to the struct called `isSet` and check that. This would be good if we didn't need firstRun for anything else - would be clearer what the code is doing.

**Comment by SwiCago** [comment]
@uronito @kayno @schotime , I just hooked up my Oscilloscope and dump the Serial packets
Here is what KUMO CLOUD sends for each mode 
FC,42,01,30,10,02,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,7B =>INFO
FC,42,01,30,10,03,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,7A =>INFO_TEMP
FC,42,01,30,10,04,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,79 => ???
FC,42,01,30,10,09,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,74 => ???
-----------------------------------------------------------------------------------
FC,42,01,30,10,- -,TP,00,pw,md,00,fs,00,00,00,00,00,00,00,tF,00,CHK  
FC,42,01,30,10,02,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,7B =>INFO
FC,41,01,30,10,01,07,00,01,02,00,00,00,00,00,00,00,00,00,A8,00,CB => DRY
FC,41,01,30,10,01,03,00,01,07,00,00,00,00,00,00,00,00,00,80,00,F2 => FAN
FC,41,01,30,10,01,07,00,01,01,00,00,00,00,00,00,00,00,00,A8,00,CC => HEAT
FC,41,01,30,10,01,07,00,01,03,00,00,00,00,00,00,00,00,00,A8,00,CA =>COOL
FC,41,01,30,10,01,05,00,00,00,00,00,00,00,00,00,00,00,00,A8,00,D0 => OFF
FC,41,01,30,10,01,08,00,00,00,00,01,00,00,00,00,00,00,00,80,00,F4 =>FANQuiet
FC,41,01,30,10,01,08,00,00,00,00,02,00,00,00,00,00,00,00,80,00,F3 =>FAN1
FC,41,01,30,10,01,08,00,00,00,00,03,00,00,00,00,00,00,00,80,00,F2 =>FAN2
FC,41,01,30,10,01,08,00,00,00,00,05,00,00,00,00,00,00,00,80,00,F0 =>FAN3
FC,41,01,30,10,01,08,00,00,00,00,06,00,00,00,00,00,00,00,80,00,EF =>FAN4
FC,41,01,30,10,01,08,00,00,00,00,00,00,00,00,00,00,00,00,80,00,F5 =>FANAuto


 Here is what I have so far...I am dumping each VANE setting now

**Comment by schotime** [comment]
Have people got this working with there own chips and kumo?

**Comment by schotime** [comment]
This was mine from last night. Connected it at midnight once figured out the wantedSettings bug.

![image](https://cloud.githubusercontent.com/assets/202923/23145737/b6f8f3e8-f824-11e6-9767-f73b6df6eb2b.png)

**Comment by kayno** [comment]
@schtime - we need to share HASS configs! Mine currently looks like: http://i.imgur.com/0VYAByU.png

**Comment by schotime** [comment]
@kayno is this a custom component?
Mine are using the standard Climate component

**Comment by SwiCago** [comment]
@schotime, @kayno it worked most cases, but I was able to find a few cases where it was wrong. What also threw update error, was any mode other than FAN. With current code I cannot HEAT or COOL. Maybe wrong temp packet. That is why I am dumping official KUMO CLOUD device. I have my logic analyzer on it pulling dumps after each change I do. I will be updating my above Comment soon, I have one more set to do and then you guys can see how it does its updates

**Comment by kayno** [comment]
@schotime I hobbled together several mqtt sensors, input_select/_slider/_boolean components, and a buck of automation rules to get that to work! did my head in, but works quite well. I will pull out just the bits related to this from my config and get it posted, when I have a chance!

**Comment by SwiCago** [comment]
@kayno @schotime @uronito , I just finished the dump, look at my above comment for info. I am thinking of replicating the KUMO cloud send style in a new branch and see if that fixes the issues some of us have.

**Comment by SwiCago** [comment]
Put my data dump in #25

**Comment by kayno** [comment]
does firstRun resolve this issue now? @uronito - can you try the latest master version please and see?

**Comment by uronito** [comment]
Hi @kayno, @swicago
I've been away for work for a couple of days and when I get back I see that you have solved all the problems (including wide vane) you and swicago. You are machines. I'm going to try the latest version and in a while I comment if it works the problem of firstrun.

Regards!!!

**Comment by uronito** [comment]
Hi @kayno, @swicago,

In your github repos in the last version , give me an error when trying upload sketch to esp8266.

exit status 1
'custom' was not declared in this scope
 Regards

**Comment by SwiCago** [comment]
@uronito, does it say what line?

**Comment by SwiCago** [comment]
@uronito grab it now. I found the issue and put in a fix

**Comment by uronito** [comment]
Thanks Swicago

**Comment by uronito** [comment]
Swicago,

continue error.

mitsubishi_heatpump_mqtt_esp8266:193: error: 'custom' was not declared in this scope

       custom.toUpperCase();

       ^
Thanks!!!!

**Comment by SwiCago** [comment]
@uronito did you pull down latest code...Takes sometimes time for github to sync right
```    if (root.containsKey("custom")) {
      String custom = root["custom"];
      custom.toUpperCase();
      char chars[40];
      custom.toCharArray(chars,40);
      byte bytes[20];
      for(int i = 0; i < 40; i+=2) {
        bytes[i] = getVal(chars[i+1])  + (getVal(chars[i]) << 4);
} ```

**Comment by uronito** [comment]
@SwiCago you're right

I'm going  to probe it

**Comment by uronito** [comment]
WOOOOOOOOWWWWWW!!!!

Works like a charm!!!!!

You're a cracks

**Comment by SwiCago** [comment]
WOOT ...  Good does this mean we can close this issue?

**Comment by uronito** [comment]
yessssss!!!!!!

The issue is solved.

:)

**Comment by SwiCago** [comment]
Great..closing. Continued discussions in #25 or open a new issue, if there is something else
